### ISSUE TITLE
Cannot change URL at run-time

### ISSUE THREAD

**Issue by benWrightSorenson** [sitespeedio/sitespeed.io#1664]
We are trying to test aspects of a Restful API without automating too much of the user interface that sits on top of it. As such, we are driving API end points to create data, then using sitespeedio to test the subsequent page load for the resources that were created. This is currently impossible, since before the resources are created we don't know the index needed to access the appropriate page. What we would like to do is provide a preScript which can update the target URL for the test.

We propose that a function "updateTargetURL" be added to the context for preScripts, which will accept a new URL. The new URL would only be used for the "loadAndWait" function, not affecting results labelling, etc.

**Comment by soulgalore** [comment]
Hey @benWrightSorenson I think it should be do-able if you create your own plugin. Look at the crawler plugin, it starts with a URL and then it [post URL messages](https://github.com/sitespeedio/sitespeed.io/blob/master/lib/plugins/crawler/index.js#L58) on the queue that other plugins will use to analyze. So for example I think you can supply a --browsertime.script  (that is a JavaScript that will run after the page is loaded (like the rest of the script we use to analyze the page), fetch the new URL, create a plugin that looks for that data from browsertime, then post that URL on the queue. @tobli do have other suggestions on how to do it? 

Best
Peter

**Comment by benWrightSorenson** [comment]
Hi @soulgalore, thanks for the speedy reply! Had a quick look at the plugin architecture, I can see how you would add additional URLs, is there a way of replacing the original URL instead?

The hack we've got locally at the moment is to have the url initialise a `targetURL` variable at the top of the `run` function, then add:
```
'updateTargetUrl': function(newUrl) {
    targetUrl = newUrl;
}
```
to the context for `preScript.run` and have the `runner.loadAndWait` function use `targetUrl` instead, format our input URLs as `http://path/with/__variable__/to/replace` and add a pre-script to swap out the __variable__ references as;
```
module.exports = {
    run: function(context) {
        var url = context.url;

        var result = [];
        var regex = /__(.+?)__/g;
        var match;

        while (match = regex.exec(url)) {
            result.push(match[1]);
        }

        result.forEach(function(match) {
            url = url.replace('__' + match + '__', URLParameters.getURLParam(match));
        });

        context.updateTargetUrl(url);
    }
};
```
where `URLParameters` is a global variable that's loaded with data by other pre-scripts. It's fairly rough-and-ready, but it lets us transform URLs as they pass through and separate out the API queries from the URL updates, while maintaining the original URL for the results labelling.

If it's possible to get the same effect using the plugins as you described then I'd be happy to go with that. Otherwise, this feels like it could evolve into a nice feature, if the regex replacement could be pulled into `engine.js`as a post-pre-script step, maybe with a function on the context to add URL variables and some validation to check for un-replaced variables to help formalise it?

Also FYI: the link to StorageManager in the docs goes to `/lib/support/storageManager.js`, think that should be `/lib/support/resultsStorage/storageManager.js`

**Comment by soulgalore** [comment]
Hi @benWrightSorenson the plugin fix will help other plugins to also use that URL, not only Browsertime. Ping @tobli what do think?

Thanks for the note about the docs, will fix it later this week!

**Comment by tobli** [comment]
NB: the incorrect docs link is here: https://www.sitespeed.io/documentation/sitespeed.io/plugins/#opencontext-options

**Comment by tobli** [comment]
Hi @benWrightSorenson!
You should consider messages to plugins in sitespeed.io, and urls in Browsertime, as immutable.

* In Sitespeed.io, the same identical message will be delivered to all plugins, to potentially act on (e.g. trigger a local browser or send a request to WebPageTest) asynchronously.
* In Browsertime, pre- and post-scripts are hooks for custom setup/teardown actions directly related to the loading of the url (e.g. logging in so the browser has a cookie set, or stopping video capture afterwards).

If you find yourself wanting to change the url as the test is running, you should look for different ways of using Sitespeed.io/Browsertime, with the above in mind. From what I understand of your use case, you're generating urls that are different every time, based on resources created via an http REST call. Why not generate the urls, and then call sitespeed with that list of urls? If you elaborate on what you're looking to get out of your tests it might become more clear.

Using sitespeed.io might not be the best tool if you're checking dynamically generated pages where the template (structure) of the page is identical, but the difference is just in the data the populating the template. A lot of performance advice and metrics will be based on the structure of your page, and thus duplicated for auto generated pages that just differ in content. Two common web performance scenarios for sitespeed.io are:
- a one time crawl of a site to detect pages with problematic content (e.g. very large images) and/or pages with problematic structure (e.g. lots of scripts or missing compression).
- repeated tests of a selected set of important pages to track improvements and regressions of performance based on changes to content and structure (e.g. html/js/css changes by developers, changes to third party includes or CMS changes by a site editor).

Other ways of using it are definitely possible, but it might be helpful to post your proposed use case before you invest to much in it, to see how the sitespeed suite of tools can be of best use to you.

**Comment by benWrightSorenson** [comment]
Hi @tobli,

Thanks for your advice. Yes, it does seem like we're working across a lot of the original design intent for sitespeedio. We chose sitespeedio after surveying a lot of the other tools available, and decided that it was closest to where we needed to be without needing to support another OS or pay out crazy amounts. We're now quite close to having the system we wanted, but it has required working around things a little.

We're primarily interested in page loading times, which means last visual change for pages based on existing resources, and loading spinners disappearing on pages where some more complex server-side computation is involved. There are a couple of long-running operations which typically follow some GUI interaction, but which we can get roughly representative run-time estimates for by creating the resource through an http request as the last pre-script, and then navigating to the page for the new resource in the test. Many of the non-functional requirements we're trying to assess for are for the same page but with different underlying data. We also need predictable concurrency, so that 5 tests can run simultaneously while 5 virtual users are accessing another part of the site, for example.

We initially used sitespeedio as you suggested for the first synchronous tests, but switched to using Amazon ECS with a different sitespeedio instance in each container, looking at a single url, to manage the concurrency element. At this point having each task generate and tear down its own data was just much neater and easier to organise than trying to create all the required data up front and then tear it all down afterwards. Also, for the long-running operations immediately following resource creation, we'd either need to create the resource DURING the test, or last thing before, making pre-scripts about the only viable option we could think of.

I think we're pretty close to having the system that we want, using the variable substitution mechanism described above. This does require us to patch browsertime, but we can carry on in that vein if this approach is too far from the direction sitespeed should be going.
